sqlliveness: fix error handling for sqlliveness sessions

aadityasondhi opened the pull request:
In a previous patch, timeouts and retries were added to the sqlliveness heartbeat loop. This introduced cases where the session could potentially go between valid, invalid, and valid state again.

This patch introduces a tenant flag for the sqlliveness instance. When enabled, the heartbeat loop is terminated if there are any errors in the session create or extend calls.

This patch also introduces an error channel which is used to propagate sqlliveness errors to the server and cli level for better visibility on the reason for process termination.

Resolves #85540

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/88837)
<!-- Reviewable:end -->

ajwerner commented:
Does this work? Doesn't stopping the stopper exit the process? Maybe we need to add some sort of closer or something.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @ajwerner)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND-4nkaEA3uN8ySXRYD):*
the commit message does not make it clear whether this patch affects only the details of how the process exits, or more than that.
___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND-55_OC4537OkXAw8O):*
I doubt that `log.Fatal` is the right thing to do; it's generally reserved for programming errors. We can do `os.Exit()` if nothing else. But hopefully we can find a more natural way to exit; how were we exiting before this patch?


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
From the code, I can see that `stopper.Stop()` is called inside `p.shutdownSQLInstance(ctx)` which results in the process exiting if the session expires. The original issue requested better logging to inform us that it exited due to session expiry and with a non-zero exit code.

An alternative to`log.Fatal()` I can think of is:
1. logging as an error
2. using `os.exit()` for the non-zero exit code

But I am not sure that is any better than `log.Fatal()`...

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-:-ND-BuNy8BjeQZjyeQRC:b-asyc18)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @ajwerner, and @andreimatei)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND-55_OC4537OkXAw8O):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I doubt that `log.Fatal` is the right thing to do; it's generally reserved for programming errors. We can do `os.Exit()` if nothing else. But hopefully we can find a more natural way to exit; how were we exiting before this patch?
</blockquote></details>

Before the patch, it would exit with exit code 0. 


<!-- Sent from Reviewable.io -->

ajwerner commented:
Can you point us at the code which was actually exiting the process?

aadityasondhi commented:
> We today don't send anything on that error chan. I think what we ought to do is hook that error chan up to this callback so that graceful shutdown happens but then we can put all the interesting information and control the exit code gracefully.

@andreimatei suggested the same to me offline. I am currently working on an implementation to use this errChan.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @ajwerner)

___
*[`-- commits` line 10 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND9GNOZDLBvQMF89WC-):*
As part of this patch, once we settle on its final form, let's take the opportunity to write a comment in the right place explaining why we terminate sql processes when the session expires. I happen to think that there are no good reasons and this is probably a bad design. Regardless, we can try to explain it through the fact that our instance id might have been taken by another, and further explain that instance ids are a somewhat scarce resource (15 bits), and a sql process needs to have a single ID throughout its lifetime (for example because it uses it as a logging tag) and so we're not willing to have an unhealthy process hold on to its id for too long, and thus we have processes terminate themselves when they're not healthy enough to maintain their session. 
___
*[`-- commits` line 9 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND90QXdBIIy75D9WBLF):*
_Start here; I've written this after other comments, and it might make some of them moot._

I don't understand [this code](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L304-L308) in the `Instance` hb loop. This branch does not call the expiration callbacks. If we like killing the process when this session expires, then how can we allow the heartbeat loop to exit without calling those callbacks?
And also the [following arm](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L309-L313), dealing with sessions deleted by others. I think similalrly we want to terminate the processes.
Also, is the [initial creation of the session](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L270) kosher? We're doing the insertion with a timeout, and potentially retrying it. The `Insert` method [says](https://github.com/cockroachdb/cockroach/blob/00af7d7089d90ccaf9d12724872881064c68e9aa/pkg/sql/sqlliveness/slstorage/slstorage.go#L394-L395):
```
// A client must never call this method with a session which was previously
// used!
```
But isn't that essentially what can happen if an insertion succeeds and the timeout also hits (or if the insertion gets an `AmbiguousCommitError`)? It looks to me like,  the second time around, the `InitPut` will start failing with `ConditionFailedError` and then be endlessly retried [by this retry loop](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L192).


If all this it is indeed broken, I think this PR is the time to fix it, perhaps in a prepended commit. We'll probably want to backport this too. The way I would do it, which ties in to the shutdown, is by officially giving the `slinstance.Instance` a "single shot" mode whereby the session doesn't go between valid/invalid/valid again. Instead, the heartbeat loop terminates at any sign of trouble, and signals some channel.

So, I'd make `slprovider.provider.Start()` return a channel that gets signaled when `slinstance.Instance.heartbeatLoop`.
I'd return that channel from [here](https://github.com/cockroachdb/cockroach/blob/d5306b044e085b04a4a2f5659085988fd80efd9a/pkg/server/server_sql.go#L1227) to [here](https://github.com/cockroachdb/cockroach/blob/d5306b044e085b04a4a2f5659085988fd80efd9a/pkg/server/tenant.go#L71), where I'd put it inside `SQLServerWrapper`, and then I'd use it out of the wrapper [here](https://github.com/cockroachdb/cockroach/blob/1af66356e77f516ede7cd56a3d7946873a590ef7/pkg/cli/mt_start_sql.go#L177)

I would move the `shutdownSQLInstance()` call [here](https://github.com/cockroachdb/cockroach/blob/4640c585e7098c53ccc69f6096e8087b6d6007fd/pkg/cli/start.go#L819); I think that's where you want it, as the last thing the server does before stopping the stopper. It looks like a bug to me that this thing wasn't being run when the process is terminated by a signal (or was it in some indirect way?)  You'll have to export the method, and I think that's fine.
With the only caller of `RegisterCallbackForSessionExpiry` gone, that method can go away.
___
*[`pkg/cli/mt_start_sql.go` line 128 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND95nrW0rSpIEb6g6Fu:-ND95nrW0rSpIEb6g6Fv:b27kv5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/cli/mt_start_sql.go#L128)):*
> ```Go
> 		ctx,
> 		stopper,
> 		errChan,
> ```

I don't like that we're plumbing this `errChan` everywhere. At the very least you need comment everywhere, and probably a better name in most places. I think it'd be more natural to have it returned by the task that will signal it - the liveness heartbeat loop.
Relatedly, I also don't like how the writing to this channel is done inside `instanceprovider.provider.initialize()`; I  think I think that this `instanceprovider` has little reason to be involved. I think it'd be a lot cleaner and much less plumbing would be needed if we lifted signaling this channel to `SQLServer.startSQLLivenessAndInstanceProviders()`. 
___
*[`pkg/cli/mt_start_sql.go` line 124 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND91qcXEGnJo5_xsaax:-ND91qcXEGnJo5_xsaay:b7a134c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f55821da24fca3d9776d294c29407f8689630460/pkg/cli/mt_start_sql.go#L124)):*
> ```Go
> 	initGEOS(ctx)
> 
> 	errChan := make(chan error, 1)
> ```

Does this channel need to be buffered? I don't think it does, and if it doesn't it's generally better to make it unbuffered as a matter of principle.
___
*[`pkg/server/server_sql.go` line 253 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8ftms5Tf_wW_l5R4I:-ND8ftms5Tf_wW_l5R4J:b1dw877) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/server/server_sql.go#L253)):*
> ```Go
> 
> 	// errChan is used to surface unrecoverable errors up to the server level to
> 	// initiate shutdown and exit with non-zero code
> ```

nit: scratch "exit with non-zero code". That's too much details about what the caller will do; the sql server doesn't care.
___
*[`pkg/server/server_sql.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8hPgJ9f2KJ4j5ak9W:-ND8hPgJ9f2KJ4j5ak9X:b3ebp1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/server/server_sql.go#L254)):*
> ```Go
> 	// errChan is used to surface unrecoverable errors up to the server level to
> 	// initiate shutdown and exit with non-zero code
> 	errChan chan error
> ```

tighten it to
```
errChan <-chan error
```
___
*[`pkg/server/testserver.go` line 901 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND91me3-M7VEb8W3gDj:-ND91me3-M7VEb8W3gDk:btwe978) ([raw file](https://github.com/cockroachdb/cockroach/blob/f55821da24fca3d9776d294c29407f8689630460/pkg/server/testserver.go#L901)):*
> ```Go
> 		baseCfg.RPCHeartbeatInterval = params.RPCHeartbeatInterval
> 	}
> 	errChan := make(chan error, 1)
> ```

ditto about buffering
___
*[`pkg/sql/sqlinstance/instanceprovider/instanceprovider.go` line 180 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8yeJyAwnk7G4CkYxl:-ND8yeJyAwnk7G4CkYxm:b-s5eurn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/sql/sqlinstance/instanceprovider/instanceprovider.go#L180)):*
> ```Go
> 		// so it can't do the shutdown (as the shutdown stops the stopper).
> 		go func() {
> 			ctx, sp := p.stopper.Tracer().StartSpanCtx(context.Background(), "instance shutdown")
> ```

nit: move the span creation and finish in `shutdownSQLInstance`
___
*[`pkg/sql/sqlinstance/instanceprovider/instanceprovider.go` line 182 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8huv_A-Ac9Z8-Vd0a:-ND8huv_A-Ac9Z8-Vd0b:b-a802tm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/sql/sqlinstance/instanceprovider/instanceprovider.go#L182)):*
> ```Go
> 			ctx, sp := p.stopper.Tracer().StartSpanCtx(context.Background(), "instance shutdown")
> 			defer sp.Finish()
> 			p.errChan <- errors.Errorf("sqlliveness: session expired")
> ```

I think you want to send on this channel after the `p.shutdownSQLInstance()` below. You don't want the process exiting to race with that call.
___
*[`pkg/sql/sqlinstance/instanceprovider/instanceprovider.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8zoma9JnByU-OaEwu:-ND8zoma9JnByU-OaEwv:b-xde3fx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/sql/sqlinstance/instanceprovider/instanceprovider.go#L203)):*
> ```Go
> 	}()
> 	select {
> 	case <-ctx.Done():
> ```

this function seems to have one caller, and the caller passes a non-cancelable context. Please get rid of this `<-ctx.Done()` in a separate commit; it's confusing.
___
*[`pkg/sql/sqlinstance/instanceprovider/instanceprovider.go` line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8iER4-KXBT7k9uH2L:-ND8iER4-KXBT7k9uH2M:bhooi0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/sql/sqlinstance/instanceprovider/instanceprovider.go#L215)):*
> ```Go
> 		log.Ops.Warningf(ctx, "could not release instance id %d", p.instanceID)
> 	}
> 	p.stopper.Stop(ctx)
> ```

get rid of this call; it should no longer be necessary. Once you do, delete the comment above that says
```
		// ... This callback runs in a stopper task,
		// so it can't do the shutdown (as the shutdown stops the stopper).
```


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Could we remove the `errChan` mechanism altogether,
and instead store the "shutdown cause" as an attribute to the `Server` struct, 
which can then be retrieved via a method in the `cli` code, thanks.

(We're going to rearchitect this code in the context of  #84700 and I would prefer not to have added complexity due to the question of who is owning the read from the channel at every time)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-:-NDT86T2AGX-UH70gZhR:b-4h14oj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @ajwerner)



<!-- Sent from Reviewable.io -->

knz reviewed (changes requested):
i'll gladly spend time to explain if there are questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-:-NDT8bqNDHuUPmSlZmPh:bk1lp0h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @ajwerner)



<!-- Sent from Reviewable.io -->

jeffswenson reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @ajwerner)

___
*[`-- commits` line 10 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND9GNOZDLBvQMF89WC-):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

As part of this patch, once we settle on its final form, let's take the opportunity to write a comment in the right place explaining why we terminate sql processes when the session expires. I happen to think that there are no good reasons and this is probably a bad design. Regardless, we can try to explain it through the fact that our instance id might have been taken by another, and further explain that instance ids are a somewhat scarce resource (15 bits), and a sql process needs to have a single ID throughout its lifetime (for example because it uses it as a logging tag) and so we're not willing to have an unhealthy process hold on to its id for too long, and thus we have processes terminate themselves when they're not healthy enough to maintain their session. 
</blockquote></details>

The fundamental purpose of sqlliveness is to bootstrap leases in the sql layer. Currently sqlliveness is used to track ownership of the sqlinstance id and jobs. @ajwerner has plans to use sqlliveness to manage sql descriptors leases. When a server's session expires we need the sql process to fence itself, since some of the leases the process owned may be claimed by another server. The simplest way to fence the server is to shut down the process. 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @ajwerner)

___
*[`-- commits` line 10 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND9GNOZDLBvQMF89WC-):*
> When a server's session expires we need the sql process to fence itself, since some of the leases the process owned may be claimed by another server.

I can't tell from your comment if you're generally defending the current behavior, or just providing facts without taking a position. In case you are defending it, I'm not sure what you mean by "fence itself". The server should not use leases that it doesn't have. It can hang in there, trying to acquire whatever leases is needs. It can also mark itself as unhealthy in order to be removed from load balancers. Exiting only to be restarted immediately by a supervisor doesn't seem great. Indeed, "kv nodes" do not exit when they can't acquire a lease, or anything like this. Grantedly, I believe there is no crash loop since, upon restart, the server blocks on creating a session.
As I understand it, simply waiting to acquire new leases doesn't quite work with the current code because it would involve getting a new instance ID (a liveness session is used to lease an instance ID, and others steal the instance ID when the session expires), but the server likes to have a stable ID throughout it's lifetime - at the very least because it copied that ID in `AmbientContexts`.  So, even if the server is able to re-acquire other types of leases, it might not be able to reacquire its instance ID.
So it seems to me that the instance ID is the reason why we exit the process. The rationalization that I've heard from @ajwerner was that instance IDs are somewhat limited in number, and so stealing them away from servers that appear to not be functioning is not entirely unreasonable.

My opinion is that is that, if we are to treat instance IDs as a precious resource, a more natural design is to burden the orchestration layer with assigning and managing them. If the orchestration layer wants to take an instance ID from one process (e.g. one that is failing to maintain its session) and use it to spawn another process, then it can do that by killing the victim explicitly (instead of relying on the victim to terminate itself).
@jeffswenson I was thinking of checking if you agree and if you're interested in doing something here, given that I know you're doing something about the session id assignments.


<!-- Sent from Reviewable.io -->

ajwerner commented:
I don't think the answer here is to involve the orchestration layer. Doing things in the orchestration layer leads to coupling I don't think anybody wants. 

If we were going back to the drawing board, I think the answer ought to be that we de-emphasize this small id which is used for unique_rowid and friends and instead give ourselves a bigger id space for use in logging and rpcs. In some ways the original sin here goes back to long ago when we assumed nodeID to be small and then used it for rowid and then assumed we needed an analog to it for the sql pods. 

All that being said, I think we've got a paradigm that mostly works and can be made to work with minimal effort. 

I'm all for redoing things that are causing lots of pain, but I don't think this is there.

andreimatei reviewed (commented):
> Could we remove the errChan mechanism altogether,
> and instead store the "shutdown cause" as an attribute to the Server struct,
> which can then be retrieved via a method in the cli code, thanks.

What I'm pushing for here is getting rid of the `stopper.Stop()` call done in the callback passed to the session. I think that's a very chaotic way to shut down the process. I'd like us to have a single caller for `stopper.Stop()`, in the cli layer (*). So it seems to me that the right design is to signal an event when the session dies, and bubble that up to the cli.
No?

(*) What I'd really want is to not shut down the process at all on session trouble, and not have this huge difference between how a sql server uses the session versus how the shared servers use it. But that's more work and people seem less convinced, so whatever.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-:-NDZexGEE-Zh9aD5L6Ei:buliqba)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @ajwerner)



<!-- Sent from Reviewable.io -->

Commit:
sqlliveness: fix error handling for sqlliveness sessions

In a previous patch, timeouts and retries were added to the sqlliveness
heartbeat loop. This introduced cases where the session could potentially
go between valid, invalid, and valid state again.

This patch introduces a tenant flag for the sqlliveness instance. When
enabled, the heartbeat loop is terminated if there are any errors in the
session create or extend calls.

This patch also introduces an error channel which is used to propogate
sqlliveness errors to the server and cli level for better visibility on
the reason for process termination.

Resolves #85540

Release note: None

aadityasondhi reviewed (commented):
Still needs some work to make the error handling work well for the non-tenant start code path. We already have a separate error channel that we listen to there. I could pass both error channels into the shutdown function as one solution. Not sure if there is something better to be done here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-:-NEaRGwg5dnrLSoc2Z0-:b-55ue2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andreimatei)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND-4nkaEA3uN8ySXRYD):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the commit message does not make it clear whether this patch affects only the details of how the process exits, or more than that.
</blockquote></details>

Fixed
___
*[`-- commits` line 9 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND90QXdBIIy75D9WBLF):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

_Start here; I've written this after other comments, and it might make some of them moot._

I don't understand [this code](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L304-L308) in the `Instance` hb loop. This branch does not call the expiration callbacks. If we like killing the process when this session expires, then how can we allow the heartbeat loop to exit without calling those callbacks?
And also the [following arm](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L309-L313), dealing with sessions deleted by others. I think similalrly we want to terminate the processes.
Also, is the [initial creation of the session](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L270) kosher? We're doing the insertion with a timeout, and potentially retrying it. The `Insert` method [says](https://github.com/cockroachdb/cockroach/blob/00af7d7089d90ccaf9d12724872881064c68e9aa/pkg/sql/sqlliveness/slstorage/slstorage.go#L394-L395):
```
// A client must never call this method with a session which was previously
// used!
```
But isn't that essentially what can happen if an insertion succeeds and the timeout also hits (or if the insertion gets an `AmbiguousCommitError`)? It looks to me like,  the second time around, the `InitPut` will start failing with `ConditionFailedError` and then be endlessly retried [by this retry loop](https://github.com/cockroachdb/cockroach/blob/9f39e5cb9364d45e80ed09c98b471da5775aef29/pkg/sql/sqlliveness/slinstance/slinstance.go#L192).


If all this it is indeed broken, I think this PR is the time to fix it, perhaps in a prepended commit. We'll probably want to backport this too. The way I would do it, which ties in to the shutdown, is by officially giving the `slinstance.Instance` a "single shot" mode whereby the session doesn't go between valid/invalid/valid again. Instead, the heartbeat loop terminates at any sign of trouble, and signals some channel.

So, I'd make `slprovider.provider.Start()` return a channel that gets signaled when `slinstance.Instance.heartbeatLoop`.
I'd return that channel from [here](https://github.com/cockroachdb/cockroach/blob/d5306b044e085b04a4a2f5659085988fd80efd9a/pkg/server/server_sql.go#L1227) to [here](https://github.com/cockroachdb/cockroach/blob/d5306b044e085b04a4a2f5659085988fd80efd9a/pkg/server/tenant.go#L71), where I'd put it inside `SQLServerWrapper`, and then I'd use it out of the wrapper [here](https://github.com/cockroachdb/cockroach/blob/1af66356e77f516ede7cd56a3d7946873a590ef7/pkg/cli/mt_start_sql.go#L177)

I would move the `shutdownSQLInstance()` call [here](https://github.com/cockroachdb/cockroach/blob/4640c585e7098c53ccc69f6096e8087b6d6007fd/pkg/cli/start.go#L819); I think that's where you want it, as the last thing the server does before stopping the stopper. It looks like a bug to me that this thing wasn't being run when the process is terminated by a signal (or was it in some indirect way?)  You'll have to export the method, and I think that's fine.
With the only caller of `RegisterCallbackForSessionExpiry` gone, that method can go away.


</blockquote></details>

I added a single shot mode for multitenant, but I am a little unsure about how to handle the error channel when it is running in non single shot mode.
___
*[`pkg/cli/mt_start_sql.go` line 128 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND95nrW0rSpIEb6g6Fu:-NEaQt_CDYhyBpqB1mRp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/cli/mt_start_sql.go#L128)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't like that we're plumbing this `errChan` everywhere. At the very least you need comment everywhere, and probably a better name in most places. I think it'd be more natural to have it returned by the task that will signal it - the liveness heartbeat loop.
Relatedly, I also don't like how the writing to this channel is done inside `instanceprovider.provider.initialize()`; I  think I think that this `instanceprovider` has little reason to be involved. I think it'd be a lot cleaner and much less plumbing would be needed if we lifted signaling this channel to `SQLServer.startSQLLivenessAndInstanceProviders()`. 

</blockquote></details>

Done.
___
*[`pkg/cli/mt_start_sql.go` line 124 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND91qcXEGnJo5_xsaax:-NEaQrQL2y9GKsr1mSY4:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f55821da24fca3d9776d294c29407f8689630460/pkg/cli/mt_start_sql.go#L124)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Does this channel need to be buffered? I don't think it does, and if it doesn't it's generally better to make it unbuffered as a matter of principle.
</blockquote></details>

Fixed
___
*[`pkg/server/server_sql.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8hPgJ9f2KJ4j5ak9W:-NEaQyE869zzC-obEg_E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/server/server_sql.go#L254)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

tighten it to
```
errChan <-chan error
```
</blockquote></details>

Done.
___
*[`pkg/server/testserver.go` line 901 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND91me3-M7VEb8W3gDj:-NEaR-PzDeXfwTX_ewea:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f55821da24fca3d9776d294c29407f8689630460/pkg/server/testserver.go#L901)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ditto about buffering
</blockquote></details>

Done.
___
*[`pkg/sql/sqlinstance/instanceprovider/instanceprovider.go` line 182 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8huv_A-Ac9Z8-Vd0a:-NEaR6Tz5SvnMDHNiyEs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/sql/sqlinstance/instanceprovider/instanceprovider.go#L182)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think you want to send on this channel after the `p.shutdownSQLInstance()` below. You don't want the process exiting to race with that call.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlinstance/instanceprovider/instanceprovider.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8zoma9JnByU-OaEwu:-NEaR8jQ8V-IVP_11kKd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/sql/sqlinstance/instanceprovider/instanceprovider.go#L203)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this function seems to have one caller, and the caller passes a non-cancelable context. Please get rid of this `<-ctx.Done()` in a separate commit; it's confusing.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlinstance/instanceprovider/instanceprovider.go` line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-ND8iER4-KXBT7k9uH2L:-NEaR9TeDrtGZvwDibUR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d0bea1bb8565bf048bb60f0c5ca3fa06f2be497/pkg/sql/sqlinstance/instanceprovider/instanceprovider.go#L215)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

get rid of this call; it should no longer be necessary. Once you do, delete the comment above that says
```
		// ... This callback runs in a stopper task,
		// so it can't do the shutdown (as the shutdown stops the stopper).
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> What I'm pushing for here is getting rid of the stopper.Stop() call done in the callback passed to the session. I think that's a very chaotic way to shut down the process.

Agreed. However, the spaghetting around of this error channel is distinctly worse. Here is the proper ranking of the possible approaches, from best to worse:

1. proper orchestration, without channel spaghetti (best)
2. using stopper.Stop in the same way as done in `(s *drainServer) maybeShutdownAfterDrain()`, `pkg/server/drain.go` (and let's extract this code into a shared function)
3. channel spaghetti as proposed here. 

>  I'd like us to have a single caller for stopper.Stop(), in the cli layer (*)

This might be a worthwhile endeavor, but is it the right PR to do this? I think there is a simpler way.

I appreciate that neither you nor Aaditya are interested to work on this particular area more (in order to find a good solution that matches point 1 above). I am OK with you not spending time on it, and leaving a TODO. (And use the `maybeShutdownAfterDrain` logic in the meantime.)

Adding the channel spaghetti is *much worse* outcome and also will take quite some time for Aaditya to figure out the kinks, which I believe is not time well spent.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88837#-:-NEaTdO13qfyT8SfVIxi:bgoqbcz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andreimatei)



<!-- Sent from Reviewable.io -->

knz commented:
There's precedent to `maybeShutdownAfterDrain` btw: in `cockroach demo`, when you run `\demo node stop` that's what it does (and the process remains running).

aadityasondhi commented:
Closing this since https://github.com/cockroachdb/cockroach/pull/90208 took care of this.
